# Builiding a Spam Filter with Naive Bayes #

In this project, we are going to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm. Our goal is to write a program that classifies new messages with an accuracy grater than 80% - so we expect that more than 80% of new messages will be classified correctly as spam or ham (non-spam).

To train the algorithm, we will use a dataset of 5,572 SMS messages that are already classified by humans. The dataset was put together by Tiago A. Almedia and José María Gómez Hidalgo, and it can be downloaded from the [The UCI machine learning repository](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection). The data collection process is described in more details on [this page](http://www.dt.fee.unicamp.br/~tiago/smsspamcollection/#composition), where you can also find some of the papers authored by Tiago A. Almeida and José María Gómez Hidalgo.

## Exploring the Dataset ##

We 'll now start by reading in the dataset.

In [1]:
import pandas as pd
sms_spam = pd.read_csv('SMSSpamCollection', sep = '\t', header = None, names = ['Label', 'SMS'])

print(sms_spam.shape)
sms_spam.head()

(5572, 2)


Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [2]:
sms_spam['Label'].value_counts(normalize = True)

ham     0.865937
spam    0.134063
Name: Label, dtype: float64

Above, we see that about 87% of the messages are ham, and the remanining 13% are spam. This sample looks representative, since in practice most messages that people receive are ham.

## Training and Test Sets ##

We are going to split the dataset into a training and a test set, where the training set accounts for 80% of the data, and the test set for the remaining 20%.

In [3]:
# Randomize the dataset
data_randomized = sms_spam.sample(frac = 1, random_state = 1)

# Calculate index for split
split_index = round(len(data_randomized) * 0.8)

# Split the randomized dataset into training and test set
training_set = data_randomized[:split_index].reset_index(drop = True)
test_set = data_randomized[split_index:].reset_index(drop = True)

print(training_set.shape)
print(test_set.shape)

(4458, 2)
(1114, 2)


We'll now analyze the percentage of ham and spam messages in the training and the test sets. We expect the percentages to be close to what we have in the full dataset, where about 87% of messages are ham, and the remainning 13% are spam.

In [4]:
training_set['Label'].value_counts(normalize = True)

ham     0.86541
spam    0.13459
Name: Label, dtype: float64

In [5]:
test_set['Label'].value_counts(normalize = True)

ham     0.868043
spam    0.131957
Name: Label, dtype: float64

The results look good. Now, we'll move on to cleaning the dataset.

## Data Cleaning ##

To calculate all the probabilities required by the algorithm, we'll first need to perform a bit of data cleaning to bring the data in a format that will allow us to extract easily all the information we need.

Essentially, we want yo bring our data to this format:

![image](format.png)

## Letter Case and Punctuations ##

We'll begin with removing all the punctuations and bring every letter to lower case.

In [6]:
# Before Cleaning
training_set.head()

Label                                                SMS
0   ham                       Yep, by the pretty sculpture
1   ham      Yes, princess. Are you going to make me moan?
2   ham                         Welp apparently he retired
3   ham                                            Havent.
4   ham  I forgot 2 ask ü all smth.. There's a card on ...

In [7]:
# After Cleaning
training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ')
training_set['SMS'] = training_set['SMS'].str.lower()
training_set.head()

/var/folders/b3/lhqcwxhj01zg5g8dk4qjg3bw0000gn/T/ipykernel_13549/125221363.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ')


Label                                                SMS
0   ham                       yep  by the pretty sculpture
1   ham      yes  princess  are you going to make me moan 
2   ham                         welp apparently he retired
3   ham                                            havent 
4   ham  i forgot 2 ask ü all smth   there s a card on ...

## Creating the Vocabulary ##

Let's now move to creating the vocabulary set, which in this context means a list of all the unique words in our training set.

In [8]:
training_set['SMS'] = training_set['SMS'].str.split()

vocabulary = []
for sms in training_set['SMS']:
    for word in sms:
        vocabulary.append(word)
        
vocabulary = list(set(vocabulary))        

In [9]:
len(vocabulary)

7783

It looks like there are 7,783 words in all our messages of our training set.

## The Final Training Set ##

We are now going to use the vocabulary set we just created to make the data transformation we want.

In [10]:
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training_set['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1

In [11]:
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()

spinout  havebeen  god  2marrow  nickey  doug  removed  global  gravy  \
0        0         0    0        0       0     0        0       0      0   
1        0         0    0        0       0     0        0       0      0   
2        0         0    0        0       0     0        0       0      0   
3        0         0    0        0       0     0        0       0      0   
4        0         0    0        0       0     0        0       0      0   

   paracetamol  ...  88600  tarot  web  veggie  advise  hop  rates  dislikes  \
0            0  ...      0      0    0       0       0    0      0         0   
1            0  ...      0      0    0       0       0    0      0         0   
2            0  ...      0      0    0       0       0    0      0         0   
3            0  ...      0      0    0       0       0    0      0         0   
4            0  ...      0      0    0       0       0    0      0         0   

   percent  crucial  
0        0        0  
1        0        0  
2        0        0  
3        0        0  
4        0        0  

[5 rows x 7783 columns]

In [12]:
training_set_clean = pd.concat([training_set, word_counts], axis = 1)
training_set_clean.head()

Label                                                SMS  spinout  havebeen  \
0   ham                  [yep, by, the, pretty, sculpture]        0         0   
1   ham  [yes, princess, are, you, going, to, make, me,...        0         0   
2   ham                    [welp, apparently, he, retired]        0         0   
3   ham                                           [havent]        0         0   
4   ham  [i, forgot, 2, ask, ü, all, smth, there, s, a,...        0         0   

   god  2marrow  nickey  doug  removed  global  ...  88600  tarot  web  \
0    0        0       0     0        0       0  ...      0      0    0   
1    0        0       0     0        0       0  ...      0      0    0   
2    0        0       0     0        0       0  ...      0      0    0   
3    0        0       0     0        0       0  ...      0      0    0   
4    0        0       0     0        0       0  ...      0      0    0   

   veggie  advise  hop  rates  dislikes  percent  crucial  
0       0       0    0      0         0        0        0  
1       0       0    0      0         0        0        0  
2       0       0    0      0         0        0        0  
3       0       0    0      0         0        0        0  
4       0       0    0      0         0        0        0  

[5 rows x 7785 columns]

## Calculating Constants First ##

We're now done with the data cleaning, and we can begin with creating the spam filter. The Naive Bayes algorithm will need to answer these two probability questions to be able to classify new messages:
$$
     P(Spam|w_{1},w_{2},...,w_{n}) = P(Spam) \cdot \prod_{n=0} ^ {n}P(w_{i}|Spam)    \tag{1}
$$

$$
     P(Ham|w_{1},w_{2},...,w_{n}) = P(Ham) \cdot \prod_{n=0} ^ {n}P(w_{i}|Ham)       \tag{2}                              
$$     

Also, to calculate the $P(w_{i}|Spam)$ and $P(w_{i}|Ham)$ inside the formulas above, we'll need to use these equations:

$$
     P(w_{i}|Spam) = \frac{N_{w_{i}|Spam} + \alpha} {N_{Spam} + \alpha \cdot N_{vocabulary}}    \tag{3}
$$

$$
     P(w_{i}|Ham) = \frac{N_{w_{i}|Ham} + \alpha} {N_{Ham} + \alpha \cdot N_{vocabulary}}       \tag{4}
$$     

Some of the terms in the four equations will have the same value for every new message. We can calculate the values of these terms once and avoid doing the computations again when a new message comes in. Below, we'll use our training set to calculate:

- P(Spam) and P(Ham)
- $N_{Spam}, N_{Ham} and N_{vocabulary}$

we'll also use Laplace Smoothing and set $\alpha$ = 1.

In [13]:
# Isolating spam and ham messages
spam_messages = training_set_clean[training_set_clean['Label'] == 'spam']
ham_messages = training_set_clean[training_set_clean['Label'] == 'ham']

# P(Spam) and P(Ham)
p_spam = len(spam_messages) / len(training_set_clean)
p_ham = len(ham_messages) / len(training_set_clean)

# N_Spam
n_words_per_spam = spam_messages['SMS'].apply(len)
n_spam = n_words_per_spam.sum()

# N_Ham
n_words_per_ham = ham_messages['SMS'].apply(len)
n_ham = n_words_per_ham.sum()

# N_vocabulary
n_vocabulary = len(vocabulary)

# Laplace Smoothing
alpha = 1

## Calculating Parameters ##

Note that we have the constant terms calculated above, we can now move on with calculating the parameters $P(w_i|Spam)$ and $P(w_i|Ham)$. Each parameter will thus be a conditional probability value associated with each word in the vocabulary.

The parameters are calculated using the formulas:

$$
    P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha} {N_{Spam} + \alpha \cdot N_{vocabulary}}    \tag{5}
$$

$$ 
    P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha} {N_{Ham} + \alpha \cdot N_{vocabulary}}       \tag{6}
$$    

In [14]:
# Initiate Parameters 
parameters_spam = {unique_word: 0 for unique_word in vocabulary}
parameters_ham = {unique_word: 0 for unique_word in vocabulary}

# Calculate Parameters
for word in vocabulary:
    n_word_given_spam = spam_messages[word].sum()     # spam_messages already defined in a cell above
    p_word_given_spam = (n_word_given_spam + alpha) / (n_spam + (alpha * n_vocabulary))
    parameters_spam[word] = p_word_given_spam
    
    n_word_given_ham = ham_messages[word].sum()       # ham_messages already classified in a cell above
    p_word_given_ham = (n_word_given_ham + alpha) / (n_ham + (alpha * n_vocabulary))
    parameters_ham[word] = p_word_given_ham

## Classifying a New Message ##

Now that we have all our parameters calculated, we can start creating the spam filter. The spam filter can be understood as a function that:

- Takes in as input a message ($w_1,w_2,...,w_n$)
- Calculate $P(Spam|w_1,w_2,...,w_n)$ and $P(Ham|w_1,w_2,...,w_n)$
- Compare the values of $P(Spam|w_1,w_2,...,w_n)$ and $P(Ham|w_1,w_2,...,w_n)$, and:
    - If $P(Ham|w_1,w_2,...,w_n)$ > $P(Spam|w_1,w_2,...,w_n)$, then the message is classified as ham.
    - If $P(Ham|w_1,w_2,...,w_n)$ < $P(Spam|w_1,w_2,...,w_n)$, then the message is classified as spam.
    - If $P(Ham|w_1,w_2,...,w_n)$ = $P(Spam|w_1,w_2,...,w_n)$, then the algorithm may request human help.

In [15]:
import re

def classify(message):
    
    message = re.sub('\W', ' ' , message)
    message = message.lower()
    message = message.split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham
    
    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
            
    print('P(Spam|message):', p_spam_given_message)
    print('P(Ham|message:', p_ham_given_message)
    
    if p_ham_given_message > p_spam_given_message:
        print('Label: Ham')
    elif p_ham_given_message < p_spam_given_message:
        print('Label: Spam')
    else:
        print('Equal probabilities, have a human classify this!')

In [16]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')

P(Spam|message): 1.3481290211300841e-25
P(Ham|message: 1.9368049028589875e-27
Label: Spam


In [17]:
classify("Sounds good, Tom, then see u there")

P(Spam|message): 2.4372375665888117e-25
P(Ham|message: 3.687530435009238e-21
Label: Ham


## Measuring the Spam Filter's Accuracy ##

The two results above look promising, but let's see how well the filter works with our test set, which has 1,114 messages.

We'll start by writing a function that returns classification labels instead of printing them.

In [21]:
def classify_test_set(message):
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham
    
    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
        
    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_ham_given_message < p_spam_given_message:
        return 'spam'
    else:
        return 'needs human classification'

Now that we have a function that returns labels instead of printing them, we can use it to create a new column in our test set.

In [22]:
test_set['predicted'] = test_set['SMS'].apply(classify_test_set)
test_set.head()

Label                                                SMS predicted
0   ham          Later i guess. I needa do mcat study too.       ham
1   ham             But i haf enuff space got like 4 mb...       ham
2  spam  Had your mobile 10 mths? Update to latest Oran...      spam
3   ham  All sounds good. Fingers . Makes it difficult ...       ham
4   ham  All done, all handed in. Don't know if mega sh...       ham

Now, we'll write a function to measure the accuracy of our spam filter to find out how well our spam filter does.

In [23]:
correct = 0
total = test_set.shape[0]

for row in test_set.iterrows():
    row = row[1]
    if row['Label'] == row['predicted']:
        correct += 1
        
print('Correct: ' , correct)
print('Incorrect: ', total - correct)
print('Accuracy: ', correct/total)

Correct:  1100
Incorrect:  14
Accuracy:  0.9874326750448833


The accuracy is close to 98.74%, which is really good. Our spam filter looked at 1,114 messages that it hasn't seen in training, and classified 1,100 correctly.